controllers/apps/operations/rebuild_instance.go

```diff
@@ -91,38 +73,92 @@ func (r rebuildInstanceOpsHandler) ActionStartedCondition(reqCtx intctrlutil.Req
 }
 
 func (r rebuildInstanceOpsHandler) Action(reqCtx intctrlutil.RequestCtx, cli client.Client, opsRes *OpsResource) error {
-	if opsRes.OpsRequest.Spec.Force {
-		return nil
-	}
 	for _, v := range opsRes.OpsRequest.Spec.RebuildFrom {
 		compStatus, ok := opsRes.Cluster.Status.Components[v.ComponentName]
 		if !ok {
 			continue
 		}
 		// check if the component has matched the `Phase` condition
-		if !slices.Contains([]appsv1alpha1.ClusterComponentPhase{appsv1alpha1.FailedClusterCompPhase,
+		if !opsRes.OpsRequest.Spec.Force && !slices.Contains([]appsv1alpha1.ClusterComponentPhase{appsv1alpha1.FailedClusterCompPhase,
 			appsv1alpha1.AbnormalClusterCompPhase, appsv1alpha1.UpdatingClusterCompPhase}, compStatus.Phase) {
 			return intctrlutil.NewFatalError(fmt.Sprintf(`the phase of component "%s" can not be %s`, v.ComponentName, compStatus.Phase))
 		}
+		var (
+			synthesizedComp *component.SynthesizedComponent
+			err             error
+			instanceNames   []string
+		)
 		for _, ins := range v.Instances {
 			targetPod := &corev1.Pod{}
 			if err := cli.Get(reqCtx.Ctx, client.ObjectKey{Name: ins.Name, Namespace: opsRes.Cluster.Namespace}, targetPod); err != nil {
 				return err
 			}
-			synthesizedComp, err := r.buildSynthesizedComponent(reqCtx, cli, opsRes.Cluster, targetPod.Labels[constant.KBAppComponentLabelKey])
+			synthesizedComp, err = r.buildSynthesizedComponent(reqCtx, cli, opsRes.Cluster, targetPod.Labels[constant.KBAppComponentLabelKey])
 			if err != nil {
 				return err
 			}
-			isAvailable, _ := r.instanceIsAvailable(synthesizedComp, targetPod, "")
-			if isAvailable {
+			isAvailable, _ := instanceIsAvailable(synthesizedComp, targetPod, "")
+			if !opsRes.OpsRequest.Spec.Force && isAvailable {
 				return intctrlutil.NewFatalError(fmt.Sprintf(`instance "%s" is availabled, can not rebuild it`, ins.Name))
 			}
+			instanceNames = append(instanceNames, ins.Name)
+		}
+		if !v.InPlace {
+			if synthesizedComp.Name != v.ComponentName {
+				return intctrlutil.NewFatalError("sharding cluster only supports to rebuild instance in place")
+			}
+			// validate when rebuilding instance with horizontal scaling
+			if err = r.validateRebuildInstanceWithHScale(reqCtx, cli, opsRes, synthesizedComp, instanceNames); err != nil {
+				return err
+			}
 		}
 	}
 	return nil
 }
 
+func (r rebuildInstanceOpsHandler) validateRebuildInstanceWithHScale(reqCtx intctrlutil.RequestCtx,
+	cli client.Client,
+	opsRes *OpsResource,
+	synthesizedComp *component.SynthesizedComponent,
+	instanceNames []string) error {
+	// rebuild instance by horizontal scaling
+	pods, err := component.ListOwnedPods(reqCtx.Ctx, cli, opsRes.Cluster.Namespace, opsRes.Cluster.Name, synthesizedComp.Name)
+	if err != nil {
+		return err
+	}
+	for _, v := range pods {
+		if slices.Contains(instanceNames, v.Name) {
```

Comment:
So if the instance to be rebuild is leader, and in place is set to false, it will fail?

Reply:
It seems unnecessary because you can use hscale to create a new pod with a follower role, and perform the rebuild on this new pod.

Reply:
> So if the instance to be rebuild is leader, and in place is set to false, it will fail?
yes.  the instance must be  unavailable and existing an available leader when inPlace is false.

Reply:
For example, an user needs to take a node offline, he may first want to rebuild all instances on that node.

Reply:
now it is supported, user can rebuild instances without health check by "force" flag.  here, only check if cluster can do horizontally scale(existing available pods).